JavaScript/JavaScriptSDK/appInsights.ts

```diff
@@ -301,7 +318,31 @@ module Microsoft.ApplicationInsights {
 
                 this.context.track(envelope);
             } catch (e) {
-                _InternalLogging.warnToConsole("trackTrace failed, trace will not be collected: " + Util.dump(e));
+                _InternalLogging.throwInternalNonUserActionable(LoggingSeverity.WARNING, "trackTrace failed, trace will not be collected: " + Util.dump(e));
+            }
+        }
+
+        /**
+        * Log a page visit time
+        * @param    pageName    Name of page
+        * @param    pageVisitDuration Duration of visit to the page in milleseconds
+        */
+        public trackPageVisitTime(pageName: string, pageUrl: string, pageVisitTime: number) {
+            var properties = { PageName: pageName, PageUrl: pageUrl };
+            this.trackMetric("PageVisitTime", pageVisitTime, 1, pageVisitTime, pageVisitTime, properties);
+        }
+
+        private autoTrackPageVisitTime(pageName: string, pageUrl: string) {
```

Comment:
I agree that moving this logic from `AppInsights` to `PageVisitTimeManager` is a good idea. In `AppInsights` this logic requires exposing the low-level `restartPageVisitTimer` method which is hard to understand here. Implemented in the `PageVisitTimeManager` class, this logic could be consumed here as a simple `trackPageVisit(pageName, url):void` method.